x509util/loginfo.go

```diff
@@ -0,0 +1,190 @@
+	}
+	return newLogInfo(log, lc)
+}
+
+// NewLogInfoOverDNS builds a LogInfo object that accesses logs via DNS, based on a log list entry.
+func NewLogInfoOverDNS(log *loglist.Log, _ *http.Client) (*LogInfo, error) {
+	if log.DNSAPIEndpoint == "" {
+		return nil, fmt.Errorf("no available DNS endpoint for log %q", log.Description)
+	}
+	dc, err := dnsclient.New(log.DNSAPIEndpoint, jsonclient.Options{PublicKeyDER: log.Key})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create DNS client for log %q: %v", log.Description, err)
+	}
+	return newLogInfo(log, dc)
+}
+
+func newLogInfo(log *loglist.Log, lc client.CheckLogClient) (*LogInfo, error) {
+	logKey, err := x509.ParsePKIXPublicKey(log.Key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse public key data for log %q: %v", log.Description, err)
+	}
+	verifier, err := ct.NewSignatureVerifier(logKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build verifier log %q: %v", log.Description, err)
+	}
+	mmd := time.Duration(log.MaximumMergeDelay) * time.Second
+	return &LogInfo{
+		Description: log.Description,
+		Client:      lc,
+		MMD:         mmd,
+		Verifier:    verifier,
+		PublicKey:   log.Key,
+	}, nil
+}
+
+// LogInfoByHash holds LogInfo objects index by the SHA-256 hash of the log's public key.
+type LogInfoByHash map[[sha256.Size]byte]*LogInfo
+
+// LogInfoByKeyHash builds a map of LogInfo objects indexed by their key hashes.
+func LogInfoByKeyHash(ll *loglist.LogList, hc *http.Client) (LogInfoByHash, error) {
+	return logInfoByKeyHash(ll, hc, NewLogInfo)
+}
+
+// LogInfoByKeyHashOverDNS builds a map of LogInfo objects (for access over DNS) indexed by their key hashes.
+func LogInfoByKeyHashOverDNS(ll *loglist.LogList, hc *http.Client) (LogInfoByHash, error) {
+	return logInfoByKeyHash(ll, hc, NewLogInfoOverDNS)
+}
+
+func logInfoByKeyHash(ll *loglist.LogList, hc *http.Client, infoFactory func(*loglist.Log, *http.Client) (*LogInfo, error)) (map[[sha256.Size]byte]*LogInfo, error) {
+	result := make(map[[sha256.Size]byte]*LogInfo)
+	for _, log := range ll.Logs {
+		h := sha256.Sum256(log.Key)
+		li, err := infoFactory(&log, hc)
+		if err != nil {
+			return nil, err
+		}
+		result[h] = li
+	}
+	return result, nil
+}
+
+// merkleVerifier is used to verify Merkle tree calculations.
+var merkleVerifier = merkletree.NewMerkleVerifier(func(data []byte) []byte {
+	hash := sha256.Sum256(data)
+	return hash[:]
+})
+
+// LastSTH returns the last STH known for the log.
+func (li *LogInfo) LastSTH() *ct.SignedTreeHead {
+	li.mu.RLock()
+	defer li.mu.RUnlock()
+	return li.lastSTH
+}
+
+// SetSTH sets the last STH known for the log.
+func (li *LogInfo) SetSTH(sth *ct.SignedTreeHead) {
+	li.mu.Lock()
+	defer li.mu.Unlock()
+	li.lastSTH = sth
+}
+
+// VerifySCTSignature checks the signature in the SCT matches the given leaf (adjusted for the
+// timestamp in the SCT) and log.
+func (li *LogInfo) VerifySCTSignature(sct ct.SignedCertificateTimestamp, leaf ct.MerkleTreeLeaf) error {
```

Comment:
Out of interest, why not pointer arguments?  A lot of the other tools return pointers, might make it easier to not have to worry about switching when passing returned values from other functions to these ones if you used pointer args in these verify functions?

Reply:
`sct` is a non-pointer because it is immediately passed to `SignatureVerifier.VerifySCTSignature`, which takes a non-pointer so I was going for consistency.  `leaf` is a non-pointer because the leaf gets its timestamp modified; it's rude to modify a passed in parameter, so given that there would need to be a copy anyway, might as well do it at argument passing time.